src/custom-element.tsx

```diff
@@ -105,6 +106,12 @@ class ShikiTextareaElement extends Element {
       adoptedStyleSheets.push(sheet(this.stylesheet))
     }
 
+    // copy event from shadowRoot to local instance
+    this.shadowRoot.addEventListener('input', e => {
+      this.value = this.shadowRoot.value
+    })
```

Comment:
Although this works, any code like this that depends on the DOM, would be cleaner in `connectedCallback`. For example:

```js
connectedCallback() {
  super.connectedCallback() // this calls `this.template()`
  
  const root = this.shadowRoot!
  const someEl = root.querySelector(...)
  // etc
}
```

Reply:
emphasis on *could be*. Depends on style. 

I have written an element where it was just props and a single `template`, f.e.

```js
class MyEl extends Element {
  @stringAttribute foo = 123
  
  template = () => {
    // .. define all refs, effects, etc, like a Solid component ...
    
    return <>...</>
  }
}
```